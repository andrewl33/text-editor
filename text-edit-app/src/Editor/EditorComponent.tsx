import * as React from "react";
// import * as Prism from 'prismjs';
import ContentEditable, { RefObject } from "react-sane-contenteditable";
import { Container, Grid } from "semantic-ui-react";
import Sidebar from "../generic/Sidebar/SidebarComponent";

import { EditorComponentProps } from "../types";
// import '../../node_modules/prismjs/themes/prism-okaidia.css';

interface EditorPrivateState {
  text: string;
}

// TODO: check https://github.com/ashleyw/react-sane-contenteditable/issues/31
// right now, the content editable does not listen to content correctly

// TODO: Update typings for react-sane-contenteditable
class EditorComponent extends React.Component<
  EditorComponentProps,
  EditorPrivateState
> {
  private delayBeforeUpdate: number;
  private myRef: RefObject<any>;
  constructor(props: EditorComponentProps) {
    super(props);

    this.myRef = React.createRef();
  }

  public render() {
    const {
      name,
      tags,
      createDate,
      isLocked,
      users,
      onRemoveTag,
      onAddTag,
      onAddUser,
      onRemoveUser,
      onNameChange,
      codeText
    } = this.props;

    return (
      <Grid style={{ marginTop: 0, height: "100%" }}>
        <Grid.Column width={13}>
          <div
            className="editor language-javascript"
            ref={this.myRef}
            style={{ height: "100%" }}
          >
            <ContentEditable
              className="code-line editor"
              content={codeText}
              editable={true}
              multiLine={true}
              onKeyDown={this.handleChange}
              style={{ padding: 0, height: "100%" }}
            />
          </div>
        </Grid.Column>
        <Grid.Column
          width={3}
          textAlign="center"
          style={{
            paddingRight: "2em",
            backgroundColor: "#3180d4",
            height: "100%"
          }}
        >
          <Container>
            <Sidebar
              name={name}
              pageType="file"
              tagList={tags}
              allTagsList={["woww", "samesame"]} // TODO:
              createDate={createDate}
              isPrivate={isLocked}
              users={users}
              onRemoveTag={onRemoveTag}
              onAddTag={onAddTag}
              onNameChange={onNameChange}
              onAddUser={onAddUser}
              onRemoveUser={onRemoveUser}
            />
          </Container>
        </Grid.Column>
      </Grid>
    );
  }

  private handleChange = (e: React.SyntheticEvent, value: string) => {
    this.setState({ text: value });
    this.props.onLocalUpdate(value);
    clearTimeout(this.delayBeforeUpdate);
    this.delay();
  };

  private delay = () => {
    // uses browser's setTimeout instead of node
    this.delayBeforeUpdate = window.setTimeout(() => {
      this.props.onBatchUpdate();
    }, 2500);
  };
}

export default EditorComponent;
