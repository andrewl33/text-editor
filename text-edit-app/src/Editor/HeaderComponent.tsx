import * as React from 'react';
import { Button, Menu } from 'semantic-ui-react';
import { EditorProps } from '../types';
import { RefObject } from 'react-sane-contenteditable';
import { ROOT_URL } from '../envConstants';
export interface HeaderProps {
  onLock?: () => void;
  onNew?: () => void;
  onShare?: () => void;
  pathname: string;
}


export class HeaderComponent extends React.Component<EditorProps, {copy: boolean}> {
  
  private urlHolder: RefObject<any>;
  
  constructor(props: EditorProps) {
    super(props);

    this.urlHolder = React.createRef();
    this.copyToClipboard = this.copyToClipboard.bind(this);

    this.state = {
      copy: false
    }
  }

  public render() {

    const { onNew, onLock, pathname } = this.props;

    return (
      <header>
        <Menu attached={true} inverted={true} primary="true" size='huge'>
          <Menu.Item name='TextEdit' />
          <Menu.Menu position='right'>
            <Menu.Item>
              <Button primary={true} onClick={onNew}>New</Button>
            </Menu.Item>
            <Menu.Item>
              <Button primary={true} onClick={onLock}>Lock</Button>
            </Menu.Item>
            <Menu.Item>
              <Button primary={true} onClick={this.copyToClipboard}>Share</Button>
            </Menu.Item>
          </Menu.Menu>
        </Menu>
        {this.state.copy ? 
          (<textarea className="url-holder" value={ROOT_URL+pathname} placeholder={"placeholder"} readOnly={true} ref={this.urlHolder}/>)
        : ''}
        
      </header>
    );
  }

  public componentDidUpdate() {
    // copy to clipboard
    // reset copy state
    if (this.state.copy) {
      this.urlHolder.current.select();
      window.document.execCommand('copy');
      this.setState({copy: false});
      this.props.onShare();
    }
  }


  
  private copyToClipboard(): void {
    // shows the textarea so it is selectable
    this.setState({copy: true});

    // the state is set back to false in componentDidUpdate()
  }
}

export default HeaderComponent;