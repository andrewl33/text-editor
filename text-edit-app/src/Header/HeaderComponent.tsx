import * as React from 'react';
import { Button, Menu } from 'semantic-ui-react';

export interface HeaderProps {
  onLock?: () => void;
  onNew?: () => void;
  onShare?: () => void;
}

const TopHeader = ({ onNew, onLock, onShare }: HeaderProps) => {
  return (
    <header>
      <Menu attached={true} inverted={true} primary="true" size='huge'>
        <Menu.Item name='TextEdit' />
        <Menu.Menu position='right'>
          <Menu.Item>
            <Button primary={true} onClick={onNew}>New</Button>
          </Menu.Item>
          <Menu.Item>
            <Button primary={true} onClick={onLock}>Lock</Button>
          </Menu.Item>
          <Menu.Item>
            <Button primary={true} onClick={onShare}>Share</Button>
          </Menu.Item>
        </Menu.Menu>
      </Menu>
    </header>
  );
}

export default TopHeader;