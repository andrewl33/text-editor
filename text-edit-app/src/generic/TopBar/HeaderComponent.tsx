import * as React from "react";
import { Button, Dropdown, Menu, Transition } from "semantic-ui-react";
import { ROOT_URL } from "../../envConstants";
import { HeaderProps } from "../../types";
import AlertComponent from "../Alert/AlertComponent";
import { PromptComponent } from "../Prompt/PromptComponent";

interface HeaderState {
  copy: boolean;
  onLockPrompt: boolean;
}

export class HeaderComponent extends React.Component<HeaderProps, HeaderState> {
  private urlHolder: React.RefObject<HTMLTextAreaElement>;
  private notificationDuration: number;

  constructor(props: HeaderProps) {
    super(props);

    this.urlHolder = React.createRef();
    this.copyToClipboard = this.copyToClipboard.bind(this);

    this.state = {
      copy: false,
      onLockPrompt: false
    };
  }

  public render() {
    const {
      onNew,
      accountName,
      loggedIn,
      pathname,
      pageName,
      isShareable = false,
      openAlert,
      alertMessage,
      getAccountCredentials,
      getPassword,
      prompt,
      onPrompt,
      onDashboard,
      onLogOut,
      onLogInPrompt,
      onClosePrompt,
      onLock,
      onHomeClick
    } = this.props;

    const alert = alertMessage !== undefined && (
      <AlertComponent open={openAlert} message={alertMessage} />
    );
    const shareButton = isShareable && (
      <Menu.Item>
        <Button primary={true} onClick={this.copyToClipboard}>
          Share
        </Button>
      </Menu.Item>
    );
    const hiddenShare = this.state.copy && (
      <textarea
        className="url-holder"
        value={ROOT_URL + pathname}
        placeholder={"placeholder"}
        readOnly={true}
        ref={this.urlHolder}
      />
    );
    const newButton = onNew && (
      <Menu.Item>
        <Button primary={true} onClick={onNew}>
          New
        </Button>
      </Menu.Item>
    );
    const lockButton = onLock && (
      <Menu.Item>
        <Button primary={true} onClick={this.handleLockRequest}>
          Lock
        </Button>
      </Menu.Item>
    );
    const accountDropDown = (
      <div>
        <Dropdown
          text={loggedIn ? accountName : "Account"}
          icon="user"
          floating={true}
          labeled={true}
          button={true}
          className="icon"
          style={{
            color: "#fff",
            backgroundColor: "#2185d0",
            fontSize: 17
          }}
        >
          {loggedIn ? (
            <Dropdown.Menu>
              <Dropdown.Item onClick={onDashboard}>Dashboard</Dropdown.Item>
              <Dropdown.Item onClick={onLogOut}>Log out</Dropdown.Item>
            </Dropdown.Menu>
          ) : (
            <Dropdown.Menu>
              <Dropdown.Item onClick={onLogInPrompt}>Log in</Dropdown.Item>
            </Dropdown.Menu>
          )}
        </Dropdown>
      </div>
    );

    return (
      <header>
        {alert}
        <Menu attached={true} inverted={true} primary="true" size="huge">
          <Menu.Item name="TextEdit" onClick={onHomeClick} />
          <Menu.Item name={pageName} />
          <Menu.Menu position="right">
            <Menu.Item>
              <Button primary={true} onClick={this.props.toGradingPage}>
                Grading
              </Button>
            </Menu.Item>
            {newButton}
            {lockButton}
            {shareButton}
            <Menu.Item>{accountDropDown}</Menu.Item>
          </Menu.Menu>
        </Menu>
        <Transition
          animation="fade down"
          unmountOnHide={true}
          visible={onPrompt || this.state.onLockPrompt}
        >
          <PromptComponent
            getAccountCredentials={getAccountCredentials}
            getPassword={getPassword}
            prompt={this.state.onLockPrompt ? "Lock" : prompt}
            onClosePrompt={onClosePrompt}
            lockItem={this.handleLockSubmit}
          />
        </Transition>
        {hiddenShare}
      </header>
    );
  }

  public componentDidUpdate = (): void => {
    // copy to clipboard
    // reset copy state
    // if (this.props.isShareable && this.state.copy) {
    if (true && this.state.copy) {
      this.urlHolder.current!.select();
      window.document.execCommand("copy");
      this.setState({ copy: false });
    }

    if (this.props.openAlert) {
      clearTimeout(this.notificationDuration);
      this.notificationDuration = window.setTimeout(this.props.onAlert, 1500);
    }
  };

  private copyToClipboard = (): void => {
    // shows the textarea so it is selectable
    this.setState({ copy: true });
    // the state is set back to false in componentDidUpdate()
  };

  private handleLockRequest = (): void => {
    this.setState({ onLockPrompt: true });
  };

  private handleLockSubmit = (password: string) => {
    this.setState({ onLockPrompt: false });
    if (this.props.onLock) {
      this.props.onLock(password);
    }
  };
}
