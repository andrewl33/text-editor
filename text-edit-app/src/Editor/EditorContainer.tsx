import * as React from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import EditorComponent from './EditorComponent';
import HeaderComponent from './HeaderComponent';
import { updateCode, changeCode, lockText, newText, shareLink, EditorAction } from './EditorAction';
import * as types from '../types';


class EditorContainer extends React.Component<types.EditorStoreState> {
  public render() {
    return (
      <div>
        <HeaderComponent {...this.props}/>
        <EditorComponent {...this.props}/>
      </div>

    );
  }
}

const mapStateToProps = (state: types.EditorStoreState) => {
  const { codeText, isLoading, isNewPage, url, hasAuth, isLocked, isSaved } = state;
  return {
    codeText, isLoading, isNewPage, url, hasAuth, isLocked, isSaved
  };
}

const mapDispatchToProps = (dispatch: Dispatch<EditorAction>) => {
  return {
    onBatchUpdate: () => dispatch(updateCode()),
    onCodeChange: () => dispatch(changeCode()),
    onLock: () => dispatch(lockText()),
    onNew: () => dispatch(newText()),
    onShare: () => dispatch(shareLink())
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditorContainer);