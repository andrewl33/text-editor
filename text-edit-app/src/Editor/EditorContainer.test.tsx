import { shallow } from "enzyme";
import * as React from "react";
import * as renderer from "react-test-renderer";
import { HeaderComponent } from "../generic/TopBar/HeaderComponent";
import EditorComponent from "./EditorComponent";
import { EditorContainer } from "./EditorContainer";

export const dispatchProps = {
  onLocalUpdate: jest.fn(),
  onBatchUpdate: jest.fn(),
  onCodeChange: jest.fn(),
  onAlert: () => jest.fn(),
  onLock: jest.fn(),
  onMount: jest.fn(),
  onNew: jest.fn(),
  onShare: jest.fn(),
  onAuthAccount: jest.fn(),
  onAuthFile: jest.fn(),
  onAddTag: jest.fn(),
  onRemoveTag: jest.fn(),
  onNameChange: jest.fn(),
  onAddUser: jest.fn(),
  onRemoveUser: jest.fn(),
  onLogInPrompt: jest.fn(),
  onClosePrompt: jest.fn(),
  onDashboard: jest.fn(),
  onLogOut: jest.fn(),
  onHomeClick: jest.fn(),
  toGradingPage: jest.fn()
};

const routerProps = {
  pathname: ""
};

const authProps = {
  accountName: "accountName",
  loggedIn: true,
  authPrompt: false
};

const editorStateProps = {
  codeText: "test",
  tags: ["tag1", "tag2"],
  isNewPage: false,
  isLocked: false,
  isSaved: false,
  openAlert: false,
  alertMessage: "none",
  filePrompt: false,
  name: "name1",
  createDate: "1-1-1",
  users: ["user1"],
  remountEditorComponent: 0
};

const staticProps = {
  pageName: "File"
};

const props = {
  ...dispatchProps,
  ...routerProps,
  ...authProps,
  ...editorStateProps,
  ...staticProps
};

const componentLoaded = shallow(
  <EditorContainer {...props} key={editorStateProps.remountEditorComponent} />
);
describe("EditorContainer", () => {
  test("renders correctly", () => {
    const component = renderer.create(
      <EditorContainer
        {...props}
        key={editorStateProps.remountEditorComponent}
      />
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("onMount called on mount", () => {
    expect(dispatchProps.onMount).toBeCalled();
  });

  test("Header Component loaded", () => {
    expect(componentLoaded.find(HeaderComponent).length).toEqual(1);
  });

  test("Editor Component loaded", () => {
    expect(componentLoaded.find(EditorComponent).length).toEqual(1);
  });
});
