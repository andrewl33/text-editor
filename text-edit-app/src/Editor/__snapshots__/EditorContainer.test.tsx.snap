// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorContainer renders correctly 1`] = `
<div
  style={
    Object {
      "height": "100%",
    }
  }
>
  <header>
    <div
      className="ui huge inverted attached menu"
      primary="true"
    >
      <a
        className="item"
        onClick={[Function]}
      >
        Text Edit
      </a>
      <div
        className="item"
        onClick={[Function]}
      >
        File
      </div>
      <div
        className="right menu"
      >
        <div
          className="item"
          onClick={[Function]}
        >
          <button
            className="ui primary button"
            onClick={[Function]}
            role="button"
          >
            New
          </button>
        </div>
        <div
          className="item"
          onClick={[Function]}
        >
          <button
            className="ui primary button"
            onClick={[Function]}
            role="button"
          >
            Lock
          </button>
        </div>
        <div
          className="item"
          onClick={[Function]}
        >
          <button
            className="ui primary button"
            onClick={[Function]}
            role="button"
          >
            Share
          </button>
        </div>
        <div
          className="item"
          onClick={[Function]}
        >
          <div>
            <div
              aria-expanded={false}
              className="ui button floating labeled dropdown icon"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              role="listbox"
              style={
                Object {
                  "backgroundColor": "#2185d0",
                  "color": "#fff",
                  "fontSize": 17,
                }
              }
              tabIndex={0}
            >
              <div
                aria-live="polite"
                className="text"
                role="alert"
              >
                accountName
              </div>
              <i
                aria-hidden="true"
                className="user icon"
                onClick={[Function]}
              />
              <div
                className="menu transition"
              >
                <div
                  className="item"
                  onClick={[Function]}
                  role="option"
                >
                  Dashboard
                </div>
                <div
                  className="item"
                  onClick={[Function]}
                  role="option"
                >
                  Log out
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </header>
  <div
    className="ui grid"
    style={
      Object {
        "height": "100%",
        "marginTop": 0,
      }
    }
  >
    <div
      className="thirteen wide column"
      style={
        Object {
          "padding": 0,
        }
      }
    >
      <div
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      >
        <div
          id="ace-editor"
          style={
            Object {
              "height": "100%",
              "position": "absolute",
              "width": "100%",
            }
          }
        />
      </div>
    </div>
    <div
      className="center aligned three wide column"
      style={
        Object {
          "backgroundColor": "#3180d4",
          "height": "100%",
          "paddingRight": "2em",
        }
      }
    >
      <div
        className="ui container"
      >
        <div
          className="ui list"
          role="list"
          style={
            Object {
              "height": "100%",
            }
          }
        >
          <div
            className="item"
            onClick={[Function]}
            role="listitem"
          >
            <div
              className="header"
            >
              Name
            </div>
            <span
              className="name"
            >
              name1
              <i
                aria-hidden="true"
                className="edit link icon"
                onClick={[Function]}
              />
            </span>
          </div>
          <div
            className="item"
            onClick={[Function]}
            role="listitem"
          >
            <div
              className="header"
            >
              Date Created
            </div>
            1-1-1
          </div>
          <div
            className="item"
            onClick={[Function]}
            role="listitem"
          >
            <div
              className="header"
            >
              Private
            </div>
            No
          </div>
          <div
            className="item users"
            onClick={[Function]}
            role="listitem"
          >
            <div
              className="header"
            >
              Users
            </div>
            <div
              className="ui label"
              onClick={[Function]}
            >
              user1
            </div>
            <span>
              <i
                aria-hidden="true"
                className="add link icon"
                onClick={[Function]}
              />
            </span>
          </div>
          <div
            className="item tags"
            onClick={[Function]}
            role="listitem"
          >
            <div
              className="header"
            >
              Tags
            </div>
            <span>
              <div
                className="ui label"
                link="true"
                onClick={[Function]}
              >
                tag1
                <i
                  aria-hidden="true"
                  className="close icon"
                />
              </div>
              <div
                className="ui label"
                link="true"
                onClick={[Function]}
              >
                tag2
                <i
                  aria-hidden="true"
                  className="close icon"
                />
              </div>
              <i
                aria-hidden="true"
                className="add link icon"
                onClick={[Function]}
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
